File: STRINGS/125. Valid Palindrome.java
//problem link: https://leetcode.com/problems/valid-palindrome/
//timeComplexity: O(n)
//spaceComplexity: o(n)

class Solution {
    public boolean isPalindrome(String s) {
        int n=s.length();
        String temp="";
        for(int i=0;i<n;i++){
            char ch=s.charAt(i);
            if(ch!=' ' && Character.isLetterOrDigit(ch)){
                temp+=Character.toLowerCase(ch);
            }
        }
        int i=0;
        int j=temp.length()-1;
        while(i<j){
            if(temp.charAt(i) != temp.charAt(j)){
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: STRINGS/125. Valid Palindrome.cpp
//problem link: https://leetcode.com/problems/valid-palindrome/
//timeComplexity: o(n)
//spaceComplexity: O(1)

class Solution {
public:
    bool isPalindrome(string s) {
        int n=s.size();
        int left=0,right=n-1;
        while(left<right){
            while(left<right && (s[left]==' ' || !isalnum(s[left])))left++;
            while(left<right && (s[right]==' '|| !isalnum(s[right])))right--;
            if(left<right && tolower(s[left])!=tolower(s[right]))return false;
            left++,right--;
        }
        return true;
    }
};